### PR TITLE
Provide HIX feedback details

### PR DESCRIPTION
**Pull request by seluianova**
### Short description
Textlab API returns some analysis details in a json response to the benchmark request.
We would like to display them in the HIX widget according [to the design](https://www.figma.com/file/6U7R7Xj4wL7sbjxKRmOG9D/CMS-Project?type=design&node-id=1013%3A646&mode=design&t=CoWs1wUEUXzZiKM4-1).

The following categories are handled in this PR:

> Schachtelsätze
> Lange Sätze
> Lange Wörter
> Sätze im Passiv
> Infinitiv-Konstruktionen
> Sätze im Nominalstil
> Sätze im Futur

image

### Proposed changes
- Add hix_feedback field to the page translation model
- Extract feedback details from the Textlab API response and save into DB

For example, this attribute represents sentences in nominal style:
```
   "countNominalStyle": [
        [
            34,
            41, 
            0
        ]
    ],
```
One array inside means that there is 1 such sentence .
For 2 sentences there will be 2 arrays:
```
   "countNominalStyle": [
        [
            34,
            41, 
            0
        ],
        [
            76,
            98, 
            1
        ],
    ],
```
etc.
- Add feedback block to hix widget html template

### Side effects
- I haven't found any but careful testing is welcome

### Resolved issues
Fixes: #2358

__________________________________________________

[Pull Request Review Guidelines](https://digitalfabrik.github.io/integreat-cms/pull-request-review-guide.html)

### CODE REVIEW

File: integreat_cms/static/src/js/analytics/hix-widget.ts
Comment (seluianova) [inline integreat_cms/static/src/js/analytics/hix-widget.ts:7]: I started to do some refactoring here to maybe improve readability a bit, because it's quite a clutter now 🙈
But decided to finish it later in order not to delay the main feature as it's high prio

File: integreat_cms/static/src/js/analytics/hix-widget.ts
Comment (seluianova) [inline integreat_cms/static/src/js/analytics/hix-widget.ts:72]: not sure if it's a best place for this kind of mapping though

File: integreat_cms/locale/de/LC_MESSAGES/django.po
Comment (JoeyStk) [inline integreat_cms/locale/de/LC_MESSAGES/django.po:5665]:
```suggestion
msgstr "Zur Überprüfung:"
```
Super nit-picky :see_no_evil:

File: integreat_cms/locale/de/LC_MESSAGES/django.po
Comment (JoeyStk) [inline integreat_cms/locale/de/LC_MESSAGES/django.po:5681]:
```suggestion
msgstr "Wörter, die zu lang sind"
```

**Review by JoeyStk — CHANGES_REQUESTED** [review]
Thanks! This is a very cool feature!
So far I couldn't find any issues. I only have two minor comments about the translation file :)

File: integreat_cms/cms/templates/hix_widget.html
Comment (deen13) [inline integreat_cms/cms/templates/hix_widget.html:77]: This seems unnecessary. I believe it would be better to look for an alternative solution if this is being used for layout.

File: integreat_cms/core/settings.py
Comment (deen13) [inline integreat_cms/core/settings.py:303]: I haven't found anything like this, but if there's a way to generate it from their documentation (maybe swagger), that would be incredibly helpful to prevent api changes from silently breaking our code.

File: integreat_cms/static/src/js/analytics/hix-widget.ts
Comment (deen13) [inline integreat_cms/static/src/js/analytics/hix-widget.ts:30]: This could be extracted into a function so that we get rid of mutable state. 

#### Calling Side
```js
const backgroundColor = calculateBackgroundColor(score)
```

#### Function Side
```js
const calculateBackgroundColor = (score, setOutdated) => {
  if (setOutdated) {
        return "rgb(16, 111, 254, 0.3)";
    } else if (score > hixThresholdGood) {
        return "rgb(74, 222, 128)";
    } else if (score > hixThresholdOk) {
        return "rgb(250, 204, 21)";
    } else {
        return "rgb(239, 68, 68)";
    }
}
```

This code introduces a pure function, `calculateBackgroundColor`, which is easy to read and also reduces complexity on the calling side for readers who don't need to understand how the color is calculated. Additionally, the `let` variable has been replaced with a `const`.

File: integreat_cms/static/src/js/analytics/hix-widget.ts
Comment (deen13) [inline integreat_cms/static/src/js/analytics/hix-widget.ts:128]: The result can be assigned directly and then also changed from `let` to `const`. ☺️ 

```
const result = await fetch(updateButton.dataset.url, ...)
    .then((response) => response.json())
    .then((json) => {
        if (json.error || json.score === null) {
            updateHixStateLabel("error");
            return [];
        } else {
            updateHixStateLabel("updated");
            initialContent = sentContent;
            return [json.score, JSON.stringify(json.feedback)];
        }
    });
```

File: integreat_cms/static/src/js/analytics/hix-widget.ts
Comment (deen13) [inline integreat_cms/static/src/js/analytics/hix-widget.ts:117]: I'd suggest to filter the feedbackCategories at the start of this function and later on use the length of that array to make the if condition more clear.

#### Sample
```
    // Filter categories with feedback
    const categoriesWithFeedback = Array.from(feedbackCategories).filter((category) => {
        const categoryName = category.getAttribute("hix-feedback-category");
        const feedbackByCategory = feedbackJson[feedbackCategoriesMapping.get(categoryName)];
        return feedbackByCategory?.length > 0;
    });

    // Loop through each category
    categoriesWithFeedback.forEach((category) => {
        const categoryCount = category.querySelector("span");
        const categoryName = category.getAttribute("hix-feedback-category");
        const feedbackByCategory = feedbackJson[feedbackCategoriesMapping.get(categoryName)];
        categoryCount.textContent = feedbackByCategory.length.toString(); // Convert to string before setting
    });

    // Show/hide feedback block based on feedback count
    const feedbackBlock = document.getElementById("hix-feedback");
    if (categoriesWithFeedback.length === 0) {
        feedbackBlock.classList.add("hidden");
    } else {
        feedbackBlock.classList.remove("hidden");
    }
```

**Review by deen13 — CHANGES_REQUESTED** [review]
Good work! 🥇

File: integreat_cms/core/settings.py
Comment (deen13) [reply integreat_cms/core/settings.py:303]: Oh my initial comment was a bit misleading: I was thinking about accessing enum types instead of string definitions.

Comment (seluianova) [reply integreat_cms/core/settings.py:303]: Do you mean moving this from settings to a separate enum class?

Comment (deen13) [reply integreat_cms/core/settings.py:303]: Oops! My initial comment might have been a bit misleading. What I meant was accessing enum types instead of string definitions directly in our code. This way, we can ensure better type safety and catch any api changes early on.

Comment (deen13) [reply integreat_cms/core/settings.py:303]: I had a quick look at their OpenAPI definition, and it seems that these values are loosely typed in their system either. Please consider this thread solved. 🙈

Comment (seluianova) [reply integreat_cms/core/settings.py:303]: Ah, right, it's hardly possible to parse their docs. We had to contact a supporting person to localize the right fields in their API response 🙈

File: integreat_cms/cms/templates/hix_widget.html
Comment (david-venhoff) [reply integreat_cms/cms/templates/hix_widget.html:77]: These spans get populated from javascript: https://github.com/digitalfabrik/integreat-cms/pull/2761/files#diff-3334c5517cdff73b29d44ccb02b3d4b1b384c84ccaa19345543ca485bcaa5a30R105

Comment (deen13) [reply integreat_cms/cms/templates/hix_widget.html:77]: That makes sense. Thanks for the explanation @david-venhoff!

File: integreat_cms/core/settings.py
Comment (david-venhoff) [inline integreat_cms/core/settings.py:302]: I know these names come from the textlab api, but could you add some comment explaining their meaning?
I don't really understand what e.g. `moreSentencesInWords` is supposed to mean.

Also I am not really sure if the settings are the best place for this. For example if we wanted to add a new category, it would not be sufficient to only update the settings. We would still have to make a new release where the mapping in `static/src/js/analytics/hix-widget.ts` is updated.
If you decide not to use the settings, maybe you could have a mapping in the textlab api client?

Roughly like this:
```python
hix_feedback_mapping = {
    "moreSentencesInClauses": "hix-count-nested-sentences",
    "moreSentencesInWords": "hix-count-long-sentences",
    "moreWordsInLetters": "hix-count-long-words",
    # ...etc
}
feedback_details = {
    cms_name: response.get(api_name, [])
    for api_name, cms_name in hix_feedback_mapping.items()
}
```
Then you could get rid of the mapping in the typescript code and directly use these values

**Review by david-venhoff — COMMENTED** [review]
Thanks :rocket: 
This looks mostly good to me, I would just suggest to not use `JsonResponse` as return type of the textlab api function

File: integreat_cms/cms/models/pages/page_translation.py
Comment (seluianova) [reply integreat_cms/cms/models/pages/page_translation.py:59]: I believe we can't have a [] state now.
It's either null (if hix is disabled or an error occured) or json with the list of categories, but their values are []

File: integreat_cms/textlab_api/textlab_api_client.py
Comment (seluianova) [reply integreat_cms/textlab_api/textlab_api_client.py:68]: Jein 😄
I have this explanation from Textlab, but tbh it didn't help me much

> At sentence level, the numbers are simply placed one below the other, i.e. start of the match, end of the match, length of the match. The length varies from the unit depending on the parameter. For nested sentences it indicates the number of sentence parts, for long sentences the number of words.
> 
> Long words only show the position of the hit and the length, in this case the number of letters.

File: integreat_cms/static/src/js/analytics/hix-widget.ts
Comment (seluianova) [reply integreat_cms/static/src/js/analytics/hix-widget.ts:128]: Then the result type will be Promise<Response>, while we need Promise<[number, string]>.

File: integreat_cms/static/src/js/analytics/hix-widget.ts
Comment (seluianova) [reply integreat_cms/static/src/js/analytics/hix-widget.ts:117]: const categoryName and const feedbackByCategory are created in 2 loops in this example

File: integreat_cms/static/src/js/analytics/hix-widget.ts
Comment (seluianova) [reply integreat_cms/static/src/js/analytics/hix-widget.ts:30]: thanks, I have applied that

File: integreat_cms/core/settings.py
Comment (seluianova) [reply integreat_cms/core/settings.py:302]: > For example if we wanted to add a new category, it would not be sufficient to only update the settings

True, but now we can remove categories without updating the code. 
No idea if this could be a useful though 🙈

Comment (seluianova) [reply integreat_cms/core/settings.py:302]: I don't have a clear opinion on where to put these categories. I'll try asking on the channel, maybe there are more opinions on this there.

File: integreat_cms/textlab_api/textlab_api_client.py
Comment (seluianova) [reply integreat_cms/textlab_api/textlab_api_client.py]: as brilliant as it's simple.

File: integreat_cms/core/settings.py
Comment (seluianova) [reply integreat_cms/core/settings.py:302]: > Then you could get rid of the mapping in the typescript code and directly use these values

Maybe that's not bad, I will think of it next week.

**Review by PeterNerlich — COMMENTED** [review]
I started a review but didn't continue and now it's no longer needed, so I'm just dumping my two cents here~~

File: integreat_cms/locale/de/LC_MESSAGES/django.po
Comment (seluianova) [reply integreat_cms/locale/de/LC_MESSAGES/django.po]: It seems you have reviewed an outdated version of the code somehow. This  has been already fixed 🙂

File: integreat_cms/locale/de/LC_MESSAGES/django.po
Comment (seluianova) [reply integreat_cms/locale/de/LC_MESSAGES/django.po]: And this is not actual as well

File: integreat_cms/static/src/js/analytics/hix-widget.ts
Comment (deen13) [reply integreat_cms/static/src/js/analytics/hix-widget.ts:129]: Sorry, here's an example that should work properly and address the following two points:

- It makes the result a const instead of a let, which removes mutability and makes the code slightly easier to read since the reader doesn't have to think about subsequent places that mutate the variables.
- It also adds error handling on the promise level, which was missing before. This could have led to a case where the server responds with anything except a status code in the range of 200 to 399, and then the client would have failed silently, if I understand correctly.

What are your thoughts on this?

```js
/**
 * Request current HIX data
 * @returns HIX score and HIX feedback
 */
const getHixData = async (): Promise<[number?, string?]> => {
    const updateButton = document.getElementById("btn-update-hix-value");
    const sentContent = getContent().trim();

    try {
        const response = await fetch(updateButton.dataset.url, {
            method: "POST",
            headers: {
                "X-CSRFToken": getCsrfToken(),
            },
            body: JSON.stringify({
                text: sentContent,
            }),
        });

        const json = await response.json();

        if (json.error || json.score === null) {
            updateHixStateLabel("error");
            return [];
        } else {
            updateHixStateLabel("updated");
            initialContent = sentContent;
            return [json.score, JSON.stringify(json.feedback)];
        }
    } catch (error) {
        updateHixStateLabel("error");
        return [];
    }
};
```

File: integreat_cms/textlab_api/textlab_api_client.py
Comment (david-venhoff) [reply integreat_cms/textlab_api/textlab_api_client.py:67]: that sounds awful!

File: integreat_cms/cms/views/utils/hix.py
Comment (seluianova) [inline integreat_cms/cms/views/utils/hix.py:98]: Need to revise this rounding change, because it's not applicable here anymore

File: integreat_cms/core/settings.py
Comment (seluianova) [reply integreat_cms/core/settings.py]: Moved the mapping to the textlab api client
Revised feedback json structure to match standard json style (property names in camel case)

File: integreat_cms/static/src/js/analytics/hix-widget.ts
Comment (seluianova) [reply integreat_cms/static/src/js/analytics/hix-widget.ts]: Also missing is the behavior of hiding/displaying relevant feedback categories (divs) depending on categoryCount.
When count is 0 - the feedback category should be hidden in the widget, otherwise - it should be displayed.

File: integreat_cms/static/src/js/analytics/hix-widget.ts
Comment (seluianova) [reply integreat_cms/static/src/js/analytics/hix-widget.ts:115]: > This could have led to a case where the server responds with anything except a status code in the range of 200 to 399

Currently the server only responds with 200, even if there is an error.
I believe this could be changed to be more in line with the http protocol, but not within the scope of this PR because we wanted to deliver HIX feedback with a high prio and refactoring existing functionality can be done later.

File: integreat_cms/static/src/js/analytics/hix-widget.ts
Comment (seluianova) [reply integreat_cms/static/src/js/analytics/hix-widget.ts]: Sorry but it also doesn't work as expected.
Hiding/displaying relevant feedback categories (divs) depending on categoryCount is missing.
When count is 0 - the feedback category should be hidden in the widget, otherwise - it should be displayed.

File: integreat_cms/textlab_api/textlab_api_client.py
Comment (david-venhoff) [inline integreat_cms/textlab_api/textlab_api_client.py:21]: Nitpick: this should probably be a normal comment, as docstrings are only used for module, class, method or function documentation (https://peps.python.org/pep-0257/#what-is-a-docstring)

**Review by david-venhoff — APPROVED** [review]
I think this is good to go after the comments from the other reviewers got resolved :+1:

**Comment by JoeyStk** [comment]
I will leave it to @deen13 to do the second review :)

**Review by deen13 — APPROVED** [review]

**Pull request merged**
